pulsar-client-cpp/lib/HTTPLookupService.cc

```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright 2016 Yahoo Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <lib/HTTPLookupService.h>
+#include <curl/curl.h>
+
+DECLARE_LOG_OBJECT()
+
+namespace pulsar {
+    const static std::string V2_PATH = "/lookup/v2/destination/";
+    const static std::string PARTITION_PATH = "/admin/persistent/";
+    const static int MAX_HTTP_REDIRECTS = 20;
+    const static std::string PARTITION_METHOD_NAME = "partitions";
+    const static int NUMBER_OF_LOOKUP_THREADS = 1;
+
+    boost::mutex HTTPLookupService::curlGlobalMutex_;
+
+    HTTPLookupService::HTTPLookupService(const std::string &lookupUrl,
+            const ClientConfiguration &clientConfiguration,
+            const AuthenticationPtr &authData)
+    : executorProvider_(boost::make_shared<ExecutorServiceProvider>(NUMBER_OF_LOOKUP_THREADS)),
+    authenticationPtr_(authData),
+    lookupTimeoutInSeconds_(clientConfiguration.getOperationTimeoutSeconds()) {
+        if (lookupUrl[lookupUrl.length() - 1] == '/') {
+            // Remove trailing '/'
+            adminUrl_ = lookupUrl.substr(0, lookupUrl.length() - 1);
+        } else {
+            adminUrl_ = lookupUrl;
+        }
+        // Once per application - https://curl.haxx.se/mail/lib-2015-11/0052.html
+        Lock lock(curlGlobalMutex_);
```

Comment:
I don't think this would work either. It just protects 2 threads from calling `curl_global_init()` at the same time, but not at different times. Also we need to keep track of how many Pulsar instances are created, since we can only call the `curl_global_cleanup()` once the last instance is destroyed.

I think the easiest option would be to encapsulate the `curl_global_init / cleanup` within a c++ object managed by a static `shared_ptr` or something like that.